File: tests/test_agent_attribution.py
#!/usr/bin/env python3
"""One-by-one Antigravity MCP attribution test for konoha agent status counters."""
import importlib
import json
import os
import shutil
import sqlite3
import subprocess
import sys
import time
import uuid

SERVER = os.path.expanduser("~/.konoha/server.py")
DB = os.path.expanduser("~/.konoha/skills.db")
STATS = os.path.expanduser("~/.konoha/agent_stats.py")
BRAIN_CLI = os.path.expanduser("~/.gemini/antigravity-cli/brain")
BRAIN_IDE = os.path.expanduser("~/.gemini/antigravity-ide/brain")

REGISTERED = frozenset(
    ["genin", "kage", "chunin", "jonin", "anbu", "tokubetsu-jonin"]
)

AGENTS = [
    ("genin", "You are a Genin scout. Log: \"[🍃 Genin] active\".", "[🍃 Genin] active. Testing."),
    ("kage", "You are the Kage. Log: \"[🌀 Kage] active\".", "[🌀 Kage] active. Testing."),
    ("chunin", "You are the Chunin intel gatherer. Log: \"[📜 Chunin] active\".", "[📜 Chunin] active. Testing."),
    ("jonin", "You are the Jonin builder. Log: \"[🛡️ Jonin] active\".", "[🛡️ Jonin] active. Testing."),
    ("anbu", "You are the Anbu agent. Log: \"[👥 Anbu] active\".", "[👥 Anbu] active. Testing."),
    (
        "tokubetsu-jonin",
        "You are the Tokubetsu Jonin scribe. Log: \"[🎯 Tokubetsu-Jonin] active\".",
        "[🎯 Tokubetsu-Jonin] active. Testing.",
    ),
]


def load_stats():
    out = subprocess.check_output([sys.executable, STATS, DB], text=True)
    return json.loads(out)


def direct_today(stats):
    return sum(
        v.get("today", 0) for k, v in stats.items() if k not in REGISTERED
    )


def last_logged_agent():
    conn = sqlite3.connect(DB)
    row = conn.execute(
        "SELECT agent FROM tool_calls WHERE tool = 'find_skill' ORDER BY id DESC LIMIT 1"
    ).fetchone()
    conn.close()
    return (row[0] or "").lower() if row else None


def mcp_find_skill_no_agent(keyword, conv_id=None):
    req_init = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "initialize",
        "params": {"protocolVersion": "2024-11-05", "clientInfo": {"name": "antigravity"}},
    }
    req_call = {
        "jsonrpc": "2.0",
        "id": 2,
        "method": "tools/call",
        "params": {
            "name": "find_skill",
            "arguments": {"keyword": keyword, "limit": 1, "compact": True},
        },
    }
    env = os.environ.copy()
    if conv_id:
        env["ANTIGRAVITY_CONVERSATION_ID"] = conv_id
    else:
        env.pop("ANTIGRAVITY_CONVERSATION_ID", None)

    payload = json.dumps(req_init) + "\n" + json.dumps(req_call) + "\n"
    proc = subprocess.run(
        [sys.executable, SERVER],
        input=payload,
        env=env,
        capture_output=True,
        text=True,
        timeout=30,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr or proc.stdout)
    if proc.stderr:
        print(f"\n--- SERVER STDERR FOR {keyword} ---", file=sys.stderr)
        print(proc.stderr, file=sys.stderr)
        print("-----------------------------------", file=sys.stderr)


def setup_brain(brain_root, prompt_text, planner_line, mtime_offset=3600):
    conv_id = f"konoha-test-{uuid.uuid4()}"
    conv_dir = os.path.join(brain_root, conv_id)
    logs = os.path.join(conv_dir, ".system_generated", "logs")
    os.makedirs(logs, exist_ok=True)
    prompt_path = os.path.join(conv_dir, "prompt.md")
    transcript_path = os.path.join(logs, "transcript.jsonl")
    with open(prompt_path, "w", encoding="utf-8") as f:
        f.write(prompt_text)
    with open(transcript_path, "w", encoding="utf-8") as f:
        f.write(
            json.dumps({"type": "PLANNER_RESPONSE", "content": planner_line}) + "\n"
        )
    now = time.time() + mtime_offset
    os.utime(prompt_path, (now, now))
    os.utime(transcript_path, (now, now))
    return conv_dir


def main():
    results = []
    keyword_base = f"konoha-agy-test-{int(time.time())}"

    for idx, (agent, prompt, planner) in enumerate(AGENTS):
        before = load_stats().get(agent, {}).get("today", 0)
        conv = setup_brain(BRAIN_CLI, prompt, planner, mtime_offset=7200 + idx * 120)
        try:
            mcp_find_skill_no_agent("jonin-skill", conv_id=os.path.basename(conv))
            logged = last_logged_agent()
            after = load_stats().get(agent, {}).get("today", 0)
            ok = logged == agent and after == before + 1
            results.append((agent, ok, logged, before, after))
            print(
                f"[{'PASS' if ok else 'FAIL'}] {agent}: "
                f"logged={logged} today {before}->{after}"
            )
        finally:
            shutil.rmtree(conv, ignore_errors=True)

    before_direct = direct_today(load_stats())
    conv = setup_brain(
        BRAIN_IDE,
        "<USER_REQUEST>\nTest orchestrator attribution\n",
        "[🌀 Orchestrator] active. Testing orchestrator.",
        mtime_offset=7200 + len(AGENTS) * 120,
    )
    try:
        mcp_find_skill_no_agent("jonin-skill", conv_id=os.path.basename(conv))
        logged = last_logged_agent()
        after_direct = direct_today(load_stats())
        ok = logged == "orchestrator" and after_direct == before_direct + 1
        results.append(
            ("orchestrator", ok, logged, before_direct, after_direct)
        )
        print(
            f"[{'PASS' if ok else 'FAIL'}] orchestrator: "
            f"logged={logged} direct today {before_direct}->{after_direct}"
        )
    finally:
        shutil.rmtree(conv, ignore_errors=True)

    failed = [r for r in results if not r[1]]
    print(f"\nPassed {len(results) - len(failed)}/{len(results)}")
    return 0 if not failed else 1


if __name__ == "__main__":
    sys.exit(main())

File: tests/test_agent_attribution.js
#!/usr/bin/env node
'use strict';

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { spawnSync } = require('child_process');
const db = require('../src/db');
const agentStats = require('../src/agent_stats');

const SERVER = path.join(os.homedir(), '.konoha', 'server.js');
const DB = path.join(os.homedir(), '.konoha', 'konoha.db');
const BRAIN_CLI = path.join(os.homedir(), '.gemini', 'antigravity-cli', 'brain');
const BRAIN_IDE = path.join(os.homedir(), '.gemini', 'antigravity-ide', 'brain');

const REGISTERED = new Set(['genin', 'kage', 'chunin', 'jonin', 'anbu', 'tokubetsu-jonin']);

const AGENTS = [
  ['genin', 'You are a Genin scout. Log: "[🍃 Genin] active".', '[🍃 Genin] active. Testing.'],
  ['kage', 'You are the Kage. Log: "[🌀 Kage] active".', '[🌀 Kage] active. Testing.'],
  ['chunin', 'You are the Chunin intel gatherer. Log: "[📜 Chunin] active".', '[📜 Chunin] active. Testing.'],
  ['jonin', 'You are the Jonin builder. Log: "[🛡️ Jonin] active".', '[🛡️ Jonin] active. Testing.'],
  ['anbu', 'You are the Anbu agent. Log: "[👥 Anbu] active".', '[👥 Anbu] active. Testing.'],
  ['tokubetsu-jonin', 'You are the Tokubetsu Jonin scribe. Log: "[🎯 Tokubetsu-Jonin] active".', '[🎯 Tokubetsu-Jonin] active. Testing.']
];

function loadStats() {
  return agentStats.computeStats(DB);
}

function directToday(stats) {
  let sum = 0;
  for (const [k, v] of Object.entries(stats)) {
    if (!REGISTERED.has(k)) {
      sum += (v.today || 0);
    }
  }
  return sum;
}

function lastLoggedAgent() {
  const conn = db.getConnection(DB);
  try {
    const row = conn.prepare("SELECT agent FROM tool_calls WHERE tool = 'find_skill' ORDER BY id DESC LIMIT 1").get();
    return (row && row.agent) ? String(row.agent).toLowerCase() : null;
  } finally {
    conn.close();
  }
}

function mcpFindSkillNoAgent(keyword, convId = null) {
  const reqInit = {
    jsonrpc: '2.0',
    id: 1,
    method: 'initialize',
    params: { protocolVersion: '2024-11-05', clientInfo: { name: 'antigravity' } }
  };
  const reqCall = {
    jsonrpc: '2.0',
    id: 2,
    method: 'tools/call',
    params: {
      name: 'find_skill',
      arguments: { keyword, limit: 1, compact: true }
    }
  };

  const env = Object.assign({}, process.env);
  if (convId) {
    env.ANTIGRAVITY_CONVERSATION_ID = convId;
  } else {
    delete env.ANTIGRAVITY_CONVERSATION_ID;
  }

  const payload = JSON.stringify(reqInit) + '\n' + JSON.stringify(reqCall) + '\n';
  const proc = spawnSync(process.execPath, [SERVER], {
    input: payload,
    env,
    encoding: 'utf8',
    timeout: 30000
  });

  if (proc.status !== 0) {
    throw new Error(proc.stderr || proc.stdout || `Server exited with ${proc.status}`);
  }
}

function setupBrain(brainRoot, promptText, plannerLine, mtimeOffset = 3600) {
  const convId = `konoha-test-${Date.now()}-${Math.floor(Math.random() * 10000)}`;
  const convDir = path.join(brainRoot, convId);
  const logs = path.join(convDir, '.system_generated', 'logs');
  fs.mkdirSync(logs, { recursive: true });

  const promptPath = path.join(convDir, 'prompt.md');
  const transcriptPath = path.join(logs, 'transcript.jsonl');

  fs.writeFileSync(promptPath, promptText, 'utf8');
  fs.writeFileSync(transcriptPath, JSON.stringify({ type: 'PLANNER_RESPONSE', content: plannerLine }) + '\n', 'utf8');

  const now = (Date.now() / 1000) + mtimeOffset;
  try {
    fs.utimesSync(promptPath, now, now);
    fs.utimesSync(transcriptPath, now, now);
  } catch (_) {}

  return convDir;
}

function main() {
  const results = [];

  for (let idx = 0; idx < AGENTS.length; idx++) {
    const [agent, prompt, planner] = AGENTS[idx];
    const before = (loadStats()[agent] || {}).today || 0;
    const conv = setupBrain(BRAIN_CLI, prompt, planner, 7200 + idx * 120);
    try {
      mcpFindSkillNoAgent('jonin-skill', path.basename(conv));
      const logged = lastLoggedAgent();
      const after = (loadStats()[agent] || {}).today || 0;
      const ok = (logged === agent && after >= before + 1);
      results.push([agent, ok, logged, before, after]);
      console.log(`[${ok ? 'PASS' : 'FAIL'}] ${agent}: logged=${logged} today ${before}->${after}`);
    } finally {
      fs.rmSync(conv, { recursive: true, force: true });
    }
  }

  const beforeDirect = directToday(loadStats());
  const conv = setupBrain(
    BRAIN_IDE,
    '<USER_REQUEST>\nTest orchestrator attribution\n',
    '[🌀 Orchestrator] active. Testing orchestrator.',
    7200 + AGENTS.length * 120
  );
  try {
    mcpFindSkillNoAgent('jonin-skill', path.basename(conv));
    const logged = lastLoggedAgent();
    const afterDirect = directToday(loadStats());
    const ok = (logged === 'orchestrator' && afterDirect >= beforeDirect + 1);
    results.push(['orchestrator', ok, logged, beforeDirect, afterDirect]);
    console.log(`[${ok ? 'PASS' : 'FAIL'}] orchestrator: logged=${logged} direct today ${beforeDirect}->${afterDirect}`);
  } finally {
    fs.rmSync(conv, { recursive: true, force: true });
  }

  const failed = results.filter(r => !r[1]);
  console.log(`\nPassed ${results.length - failed.length}/${results.length}`);
  if (failed.length > 0) process.exit(1);
}

main();
